test/foundry/TokenTransferrer.t.sol

```diff
@@ -0,0 +1,274 @@
+        TestERC20Revert tokenRevert;
+        tokenRevert = new TestERC20Revert();
+        vm.label(address(tokenRevert), "tokenRevert");
+
+        revertTransfer[0] = ConduitTransfer(
+            ConduitItemType.ERC20,
+            address(tokenRevert),
+            address(alice),
+            address(bob),
+            0,
+            1
+        );
+
+        // I thought I'd be getting the TokenTransferGenericFailure for both
+        // this test and the notOkTransfer test below, but I'm not.
+        // vm.expectRevert(
+        //     abi.encodeWithSignature(
+        //         "TokenTransferGenericFailure(address, address, address, uint256, uint256)",
+        //         address(tokenRevert),
+        //         address(alice),
+        //         address(bob),
+        //         0,
+        //         1
+        //     )
+        // );
+        // No bubbling up in the reference contract.
+        vm.expectRevert();
+        test(
+            this.execute,
+            Context(referenceConduit, revertTransfer, noCodeBatchTransfer)
+        );
+        vm.expectRevert("Some ERC20 revert message");
+        test(
+            this.execute,
+            Context(conduit, revertTransfer, noCodeBatchTransfer)
+        );
+
+        // Test the generic failure case where the token contract returns not OK but does not revert.
+        ConduitTransfer[] memory notOkTransfer;
+        notOkTransfer = new ConduitTransfer[](1);
+
+        TestERC20NotOk tokenNotOk;
+        tokenNotOk = new TestERC20NotOk();
+        vm.label(address(tokenNotOk), "tokenNotOk");
+
+        vm.startPrank(alice);
+        tokenNotOk.mint(alice, 100);
+        tokenNotOk.approve(address(consideration), uint256(100));
+        tokenNotOk.approve(address(referenceConsideration), uint256(100));
+        tokenNotOk.approve(address(conduit), uint256(100));
+        tokenNotOk.approve(address(referenceConduit), uint256(100));
+        vm.stopPrank();
+
+        notOkTransfer[0] = ConduitTransfer(
+            ConduitItemType.ERC20,
+            address(tokenNotOk),
+            address(alice),
+            address(bob),
+            0,
+            1
+        );
+
+        vm.expectRevert();
+        test(
+            this.execute,
+            Context(referenceConduit, notOkTransfer, noCodeBatchTransfer)
+        );
+        vm.expectRevert();
+        test(
+            this.execute,
+            Context(conduit, notOkTransfer, noCodeBatchTransfer)
+        );
+
+        // Test the ERC721 revert case.
+        TestERC721Revert nonfungibleTokenRevert;
+        nonfungibleTokenRevert = new TestERC721Revert();
+        vm.label(address(nonfungibleTokenRevert), "nonfungibleTokenRevert");
+
+        revertTransfer[0] = ConduitTransfer(
+            ConduitItemType.ERC721,
+            address(nonfungibleTokenRevert),
+            address(alice),
+            address(bob),
+            0,
+            1
+        );
+
+        // No bubbling up in the reference contract.
```

Comment:
if the reason we're explicitly calling `test` on different implementations rather than passing them via context is because they revert differently - we could add a state variable `bool reference` (or similar) to conditionally expect one error or another without duplicating the calls